### PR TITLE
PKCS#11  Ed25519

### PR DESCRIPTION
**Pull request by varder**
ta: pkcs11: Add Ed25519 support
Add functionality to generate, import keys, sign/verify for
Ed25519, Ed25519ctx and Ed25519ph.

### CODE REVIEW

File: ta/pkcs11/src/processing_ec.c [hunk 1]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	rc = serialargs_get_ptr(&args, &ctx_data, ctx_len);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&args))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	proc->extra_ctx = TEE_Malloc(sizeof(struct eddsa_processing_ctx) +
+								 ctx_len,
+								 TEE_USER_MEM_HINT_NO_FILL_ZERO);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_ec.c:821]:
```patch
 	proc->extra_ctx = TEE_Malloc(sizeof(struct eddsa_processing_ctx) +
-								 ctx_len,
-								 TEE_USER_MEM_HINT_NO_FILL_ZERO);
+				      ctx_len, TEE_USER_MEM_HINT_NO_FILL_ZERO);

File: ta/pkcs11/src/token_capabilities.c [hunk 2]
```diff
@@ -390,9 +395,15 @@ void pkcs11_mechanism_supported_key_sizes(uint32_t proc_id,
 	case PKCS11_CKM_ECDSA_SHA384:
 	case PKCS11_CKM_ECDSA_SHA512:
 	case PKCS11_CKM_ECDH1_DERIVE:
+	case PKCS11_CKFM_GENERATE_KEY_PAIR:
```
Comment (etienne-lms) [inline ta/pkcs11/src/token_capabilities.c:398]: discard this change

File: ta/pkcs11/src/processing_ec.c [hunk 3]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+
+	return rc;
+}
+
+enum pkcs11_rc
+pkcs2tee_proc_params_eddsa(struct active_processing *proc,
+				  struct pkcs11_attribute_head *proc_params)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	uint32_t flag = 0, ctx_len = 0;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_ec.c:797]: 1 variable definition per line

**Review by etienne-lms — COMMENTED** [review]
Thanks for the port to pkcs11 TA.
I've pointed the coding style issues, mostly indentation. See also few other comments.
That said, the overall looks good but i'll need another pass.

File: ta/pkcs11/include/pkcs11_ta.h [hunk 4]
```diff
@@ -1169,6 +1169,8 @@ enum pkcs11_key_type {
 	PKCS11_CKK_DSA				= 0x001,
 	PKCS11_CKK_DH				= 0x002,
 	PKCS11_CKK_EC				= 0x003,
+    PKCS11_CKK_EC_EDWARDS		= 0x040,
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/include/pkcs11_ta.h]: fixed

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 5]
```diff
@@ -1055,6 +1072,14 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 			goto out;
 		}
 		break;
+	case PKCS11_CKM_EC_EDWARDS_KEY_PAIR_GEN:
+		if ((get_class(temp) != PKCS11_CKO_PUBLIC_KEY &&
+			 get_class(temp) != PKCS11_CKO_PRIVATE_KEY) ||
+			get_key_type(temp) != PKCS11_CKK_EC_EDWARDS) {
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 6]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+
+	rc = serialargs_get_ptr(&args, &ctx_data, ctx_len);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&args))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	proc->extra_ctx = TEE_Malloc(sizeof(*ctx) + ctx_len,
+					     TEE_USER_MEM_HINT_NO_FILL_ZERO);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_ec.c:821]: indentation

File: ta/pkcs11/src/processing_asymm.c [hunk 7]
```diff
@@ -701,10 +741,10 @@ enum pkcs11_rc step_asymm_operation(struct pkcs11_session *session,
 
 		case PKCS11_FUNCTION_SIGN:
 			res = TEE_AsymmetricSignDigest(proc->tee_op_handle,
-						       tee_attrs,
-						       tee_attrs_count,
-						       in_buf, in_size,
-						       out_buf, &out_size);
+						tee_attrs,
+						tee_attrs_count,
+						hash_buf, hash_size,
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:746]: discard this change. The original code was good:
```c
			res = TEE_AsymmetricSignDigest(proc->tee_op_handle,
						       tee_attrs,
						       tee_attrs_count,
						       in_buf, in_size,
						       out_buf, &out_size);
```

**Review by etienne-lms — COMMENTED** [review]
few remaining comments on the pkcs11 TA part.

**Comment by jforissier** [comment]
@varder when you address Etienne's comments please also rebase onto master which now has the "proper" Ed25519 implementation, thanks!

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 8]
```diff
@@ -1742,6 +1777,22 @@ check_parent_attrs_against_processing(enum pkcs11_mechanism_id proc_id,
 		}
 		break;
 
+	case PKCS11_CKM_EDDSA:
+		if (key_type != PKCS11_CKK_EC_EDWARDS) {
+			EMSG("Invalid key %s for mechanism %s",
+				id2str_type(key_type, key_class),
+				id2str_proc(proc_id));
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: fixed

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 9]
```diff
@@ -1742,6 +1777,22 @@ check_parent_attrs_against_processing(enum pkcs11_mechanism_id proc_id,
...
+			return PKCS11_CKR_KEY_TYPE_INCONSISTENT;
+		}
+		if (key_class != PKCS11_CKO_PUBLIC_KEY &&
+			key_class != PKCS11_CKO_PRIVATE_KEY) {
+			EMSG("Invalid key class for mechanism %s",
+				id2str_proc(proc_id));
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 10]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	if (rc)
+		return rc;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_EC_POINT,
+								tee_obj, TEE_ATTR_ED25519_PUBLIC_VALUE);
+	if (rc)
+		return rc;
+
+	rc = tee2pkcs_add_attribute(pub_head, PKCS11_CKA_EC_POINT,
+								tee_obj, TEE_ATTR_ED25519_PUBLIC_VALUE);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 11]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	if (rc)
+		return rc;
+
+	out:
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_ec.c (hunk 1 above)
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: Fixed indentation,
Is it Ok? 
sizeof(*ctx) + ctx_len,

File: ta/pkcs11/src/token_capabilities.c (hunk 2 above)
Comment (varder) [reply ta/pkcs11/src/token_capabilities.c]: Indeed, thanks for pointing out

File: ta/pkcs11/src/processing_ec.c (hunk 3 above)
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing.c [hunk 12]
```diff
@@ -503,6 +504,9 @@ enum pkcs11_rc entry_generate_key_pair(struct pkcs11_client *client,
 
 	/* Generate key pair */
 	switch (proc_params->id) {
+    case PKCS11_CKM_EC_EDWARDS_KEY_PAIR_GEN:
+        rc = generate_eddsa_keys(proc_params, &pub_head, &priv_head);
+        break;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing.c]: fixed

File: ta/pkcs11/src/processing.h [hunk 13]
```diff
@@ -45,6 +45,20 @@ struct rsa_oaep_processing_ctx {
...
 };
 
+/**
+ * EDDSA processing context
+ *
+ * @flag: Prehash flag
+ * @ctx: Context data
+ * @ctx_len: Length of the context data
+ */
+
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing.h]: fixed

File: ta/pkcs11/src/processing.h [hunk 14]
```diff
@@ -45,6 +45,20 @@ struct rsa_oaep_processing_ctx {
 	uint8_t source_data[];
 };
 
+/**
+ * EDDSA processing context
+ *
+ * @flag: Prehash flag
+ * @ctx: Context data
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing.h]: fixed

File: ta/pkcs11/src/processing.h [hunk 15]
```diff
@@ -164,6 +182,10 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
 				struct obj_attrs **pub_head,
 				struct obj_attrs **priv_head);
 
+enum pkcs11_rc generate_eddsa_keys(struct pkcs11_attribute_head *proc_params,
+                struct obj_attrs **pub_head,
+                struct obj_attrs **priv_head);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing.h]: fixed

File: ta/pkcs11/src/processing.h [hunk 16]
```diff
@@ -187,6 +209,10 @@ enum pkcs11_rc
 pkcs2tee_proc_params_rsa_oaep(struct active_processing *proc,
 			      struct pkcs11_attribute_head *proc_params);
 
+enum pkcs11_rc
+pkcs2tee_proc_params_eddsa(struct active_processing *proc,
+                  struct pkcs11_attribute_head *proc_params);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing.h]: fixed

File: ta/pkcs11/src/processing_asymm.c [hunk 17]
```diff
@@ -492,6 +507,29 @@ enum pkcs11_rc step_asymm_operation(struct pkcs11_session *session,
 				       rsa_pss_ctx->salt_len, 0);
 		tee_attrs_count++;
 		break;
+	case PKCS11_CKM_EDDSA:
+		eddsa_ctx = proc->extra_ctx;
+
+		tee_attrs = TEE_Malloc(2 * sizeof(TEE_Attribute),
+							   TEE_USER_MEM_HINT_NO_FILL_ZERO);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: fixed

File: ta/pkcs11/src/processing_asymm.c [hunk 18]
```diff
@@ -492,6 +507,29 @@ enum pkcs11_rc step_asymm_operation(struct pkcs11_session *session,
...
+		if (eddsa_ctx->flag){
+			TEE_InitValueAttribute(&tee_attrs[tee_attrs_count],
+								   TEE_ATTR_EDDSA_PREHASH, 0, 0);
+			tee_attrs_count++;
+		}
+
+		if (eddsa_ctx->ctx_len > 0){
+			TEE_InitRefAttribute(&tee_attrs[tee_attrs_count],
+					   TEE_ATTR_EDDSA_CTX, eddsa_ctx->ctx,
+					   eddsa_ctx->ctx_len);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 19]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	switch (get_class(obj->attributes)) {
+	case PKCS11_CKO_PUBLIC_KEY:
+		attrs = TEE_Malloc(1 * sizeof(TEE_Attribute),
+				   TEE_USER_MEM_HINT_NO_FILL_ZERO);
+		if (!attrs)
+			return PKCS11_CKR_DEVICE_MEMORY;
+
+		if (pkcs2tee_load_attr(&attrs[count],
+						TEE_ATTR_ED25519_PUBLIC_VALUE,
+						obj, PKCS11_CKA_EC_POINT))
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 20]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+{
+	TEE_Attribute *attrs = NULL;
+	size_t count = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
+	assert(get_key_type(obj->attributes) == PKCS11_CKK_EC_EDWARDS);
+
+	switch (get_class(obj->attributes)) {
+	case PKCS11_CKO_PUBLIC_KEY:
+		attrs = TEE_Malloc(1 * sizeof(TEE_Attribute),
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 21]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	} else {
+		TEE_Free(attrs);
+	}
+
+	return rc;
+}
+
+enum pkcs11_rc generate_eddsa_keys(struct pkcs11_attribute_head *proc_params,
+								struct obj_attrs **pub_head,
+								struct obj_attrs **priv_head)
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 22]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	TEE_ObjectHandle tee_obj = TEE_HANDLE_NULL;
+	TEE_Attribute tee_key_attr[1] = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!proc_params || !*pub_head || !*priv_head)
+		return PKCS11_CKR_TEMPLATE_INCONSISTENT;
+
+	if (remove_empty_attribute(pub_head, PKCS11_CKA_EC_POINT) ||
+		remove_empty_attribute(priv_head, PKCS11_CKA_VALUE) ||
+		remove_empty_attribute(priv_head, PKCS11_CKA_EC_PARAMS)) {
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_ec.c [hunk 23]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+	tee_size = ec_params2tee_keysize(a_ptr, a_size);
+	if (!tee_size)
+		return PKCS11_CKR_ATTRIBUTE_TYPE_INVALID;
+
+	tee_curve = ec_params2tee_curve(a_ptr, a_size);
+
+	TEE_InitValueAttribute(tee_key_attr, TEE_ATTR_ECC_CURVE, tee_curve, 1);
+
+	res = TEE_AllocateTransientObject(TEE_TYPE_ED25519_KEYPAIR, tee_size,
+									  &tee_obj);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/token_capabilities.c [hunk 24]
```diff
@@ -393,6 +400,11 @@ void pkcs11_mechanism_supported_key_sizes(uint32_t proc_id,
 		*min_key_size = 160;	/* in bits */
 		*max_key_size = 521;	/* in bits */
 		break;
+	case PKCS11_CKM_EC_EDWARDS_KEY_PAIR_GEN:
+	case PKCS11_CKM_EDDSA:
+		*min_key_size = 256;	/* in bits */
```
Comment (etienne-lms) [inline ta/pkcs11/src/token_capabilities.c:405]: Spec mentions 255 while key effective bit size is 256.
Maybe a word here? `/* in bits, v3.1 falsly state 255 */`

File: ta/pkcs11/src/processing_ec.c [hunk 25]
```diff
@@ -624,6 +641,195 @@ enum pkcs11_rc generate_ec_keys(struct pkcs11_attribute_head *proc_params,
...
+
+	/* Private key needs the same EC_PARAMS as used by the public key */
+	rc = add_attribute(priv_head, PKCS11_CKA_EC_PARAMS, a_ptr, a_size);
+	if (rc)
+		goto out;
+
+	rc = tee2pkcs_add_attribute(priv_head, PKCS11_CKA_VALUE,
+				    tee_obj, TEE_ATTR_ED25519_PRIVATE_VALUE);
+	if (rc)
+		return rc;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

File: ta/pkcs11/src/processing_asymm.c [hunk 26]
```diff
@@ -701,10 +741,10 @@ enum pkcs11_rc step_asymm_operation(struct pkcs11_session *session,
...
 		case PKCS11_FUNCTION_SIGN:
 			res = TEE_AsymmetricSignDigest(proc->tee_op_handle,
-						       tee_attrs,
-						       tee_attrs_count,
-						       in_buf, in_size,
-						       out_buf, &out_size);
+						tee_attrs,
+						tee_attrs_count,
+						in_buf, in_size,
+						out_buf, &out_size);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: fixed

File: ta/pkcs11/src/token_capabilities.c (hunk 24 above)
Comment (varder) [reply ta/pkcs11/src/token_capabilities.c:405]: According to rfc8032
https://www.rfc-editor.org/rfc/rfc8032#section-5.1.5
Key Generation
The private key is 32 octets (256 bits, corresponding to b) of
cryptographically secure random data.

I guess it is a typo in Specs,

File: ta/pkcs11/src/processing_ec.c [hunk 27]
```diff
@@ -41,6 +42,24 @@ static const uint8_t secp521r1_name_der[] = {
...
+	'2', '5', '5', '1', '9',
+};
+
+static const uint8_t ed25519ph_name_der[] = {
+	0x06, 0x03, 0x2B, 0x65, 0x72,
+};
+
+static const uint8_t ed25519ph_oid_der[] = {
+	0x13, 0x0C, 'e', 'd', 'w', 'a', 'r', 'd', 's',
+	'2', '5', '5', '1', '9',
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_ec.c:60]: missing `'p', 'h'` ?

File: ta/pkcs11/src/processing_ec.c [hunk 28]
```diff
@@ -41,6 +42,24 @@ static const uint8_t secp521r1_name_der[] = {
 	0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x23,
 };
 
+static const uint8_t ed25519_name_der[] = {
+	0x06, 0x03, 0x2B, 0x65, 0x70,
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c:46]: For the existing algos the struct values
struct supported_ecc_curve
can be generated using dump_ec_curve_params.sh
with the commands:
openssl ecparam -name ${EC_CURVE} -param_enc explicit
But it looks like for Ed25519 with Openssl impossible to get the parameters.
I would like to generate it automatically, but i could find the way.

I have looked at some existing implementation,
and it looks like, 1.3.101.112 is used in this context
https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/crypto/BotanEDPrivateKey.cpp#L54
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15-algo.c#L451

Also, looking at the RFC
https://www.rfc-editor.org/rfc/rfc8410
10.3.  Examples of Ed25519 Private Key
OBJECT IDENTIFIER
Ed 25519 signature algorithm { 1 3 101 112 }

https://www.rfc-editor.org/rfc/rfc8420.html
A.1.  ASN.1 Object for Ed25519
   id-Ed25519 OBJECT IDENTIFIER ::= { 1.3.101.112 }
	Binary encoding: 3005 0603 2B65 70

[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: For the existing algos the struct values
struct supported_ecc_curve
can be generated using dump_ec_curve_params.sh
with the commands:
openssl ecparam -name ${EC_CURVE} -param_enc explicit
But it looks like for Ed25519 with Openssl impossible to get the parameters.
I would like to generate it automatically, but i could find the way.

I have looked at some existing implementation,
and it looks like, 1.3.101.112 is used in this context
https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/crypto/BotanEDPrivateKey.cpp#L54
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15-algo.c#L451

**Comment by varder** [comment]
@etienne-lms
Can I squash the fixups?

File: ta/pkcs11/src/processing_ec.c [hunk 29]
```diff
@@ -244,6 +245,21 @@ static const uint8_t secp521r1_oid_der[] = {
 	0x91, 0x38, 0x64, 0x09, 0x02, 0x01, 0x01,
 };
 
+/* Edwards curves may be specified in two flavours:
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_ec.c:248]:
```c
/*
 * Edwards curves may be specified in two flavours: 
 ```

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` (minor inline comment style issue)

File: ta/pkcs11/src/processing_ec.c (hunk 29 above)
Comment (varder) [reply ta/pkcs11/src/processing_ec.c]: fixed

**Pull request merged**